Issue #3913 Fix races in session request reference counting

janbartel opened the pull request:
#3913
Fixed races in the reference counting used by sessioncache to work out when it can evict a session. As part of fixing races, also discovered a couple of other problem cases to do with complex server-side forwarding chains to other contexts, and also problems with cleaning up sessions on async dispatches. The code is now a lot simpler and clearer in the lifecycle and reference counts for sessions.

olamy commented on jetty-server/src/main/java/org/eclipse/jetty/server/session/AbstractSessionCache.java:
```diff
@@ -111,12 +111,13 @@
      * @return the Session object matching the id
      */
-    public abstract Session doGet(String id);
+    protected abstract Session doGet(String id);
```
changing visibility in minor release? (maybe in 10?)

janbartel replied:
Yes, bit naughty I know. These methods should have been protected from the start. I figure that as they are only called by the session cache itself, then there should be no incompatibilities with this change.

olamy commented on jetty-server/src/main/java/org/eclipse/jetty/server/session/AbstractSessionCache.java:
```diff
@@ -125,7 +126,7 @@
      * @return null if the session wasn't already in the map, or the existing entry otherwise
      */
-    public abstract Session doPutIfAbsent(String id, Session session);
+    protected abstract Session doPutIfAbsent(String id, Session session);
```
changing visibility in minor release? (maybe in 10?)

janbartel replied:
See previous comment. I guess if I absolutely had to, I could leave the viz changes until 10 ...

olamy commented on jetty-server/src/main/java/org/eclipse/jetty/server/session/AbstractSessionCache.java:
```diff
@@ -135,7 +136,7 @@
      * @return true if replacement was done
      */
-    public abstract boolean doReplace(String id, Session oldValue, Session newValue);
+    protected abstract boolean doReplace(String id, Session oldValue, Session newValue);
```
changing visibility in minor release? (maybe in 10?)

janbartel replied:
As per previous comment.

sbordet commented on jetty-server/src/main/java/org/eclipse/jetty/server/Request.java:
```diff
@@ -367,6 +368,39 @@ public void addEventListener(final EventListener listener)
+        Session session = Session.class.cast(s);
+        if (LOG.isDebugEnabled())
+            LOG.debug("Request {} leaving session {}", this, session);
+        session.getSessionHandler().complete(session);
```
This strikes me as asymmetric: why the "enter" event does not need to tell the `SessionHandler` but adds the session to the list, but the "leave" event needs to tell to the `SessionHandler` and _not_ remove the session from the list?
I mean I can see the list is cleared out in `recycle()`, but often asymmetries of this kind are indication of a problem.

janbartel replied:
The SessionHandler tells the Request that it has entered the Session. The Request knows when it is completing, and tells the SessionHandler that it is leaving the Session. The Request could iterate over the list and remove each Session as it calls its SessionHandler with it. I just didn't bother because the next thing that happens is that the Request gets recycled.

sbordet commented on jetty-server/src/main/java/org/eclipse/jetty/server/Request.java:
```diff
@@ -1466,6 +1500,46 @@ public String changeSessionId()
+     */
+    public void onCompleted()
+    {
+        if (_sessions != null && _sessions.size() > 0)
```
No need to test for `_session.size > 0`, the iterator below will be easily eliminated by the JIT.

sbordet commented on jetty-server/src/main/java/org/eclipse/jetty/server/Request.java:
```diff
@@ -1466,6 +1500,46 @@ public String changeSessionId()
+        if (_sessions != null && _sessions.size() > 0)
+        {
+            for (HttpSession s:_sessions)
+                leaveSession(s);
```
Reformat this `for` following the rules: no spaces around `:` and braces are mandatory.

sbordet commented on jetty-server/src/main/java/org/eclipse/jetty/server/Request.java:
```diff
@@ -1466,6 +1500,46 @@ public String changeSessionId()
+        
+        for (HttpSession s:_sessions)
+        {
+            Session ss =  Session.class.cast(s);
```
Use simple cast.

sbordet commented on jetty-server/src/main/java/org/eclipse/jetty/server/Request.java:
```diff
@@ -1466,6 +1500,46 @@ public String changeSessionId()
+            if (sessionHandler == ss.getSessionHandler())
+            {
+                session = s;
+                break;
```
Simplify all this and just `return` from here.

sbordet commented:
@janbartel in order to review this PR I need a clear description of what the problem is. 
The issue mentions a nasty race but does not tell exactly what/where.

Also, I need a clear description of what the solution is to solve that race.

sbordet commented on jetty-server/src/main/java/org/eclipse/jetty/server/session/Session.java:
```diff
@@ -212,6 +211,8 @@ public Session(SessionHandler handler, SessionData data)
         _sessionInactivityTimer = new SessionInactivityTimer();
     }
+    
+    
```
Unnecessary blank lines.

janbartel replied:
But I like them

sbordet commented on jetty-server/src/main/java/org/eclipse/jetty/server/session/Session.java:
```diff
@@ -220,10 +221,12 @@ public Session(SessionHandler handler, SessionData data)
         try (Lock lock = _lock.lock())
         {
-            return _requests;
+            r = _requests;
```
Just `return` from here.

sbordet commented on jetty-server/src/main/java/org/eclipse/jetty/server/session/Session.java:
```diff
@@ -239,11 +242,24 @@ protected void cookieSet()
         }
     }
 
+    protected void use ()
```
Remove spaces after method name.

sbordet commented on jetty-server/src/main/java/org/eclipse/jetty/server/session/AbstractSessionCache.java:
```diff
@@ -447,7 +473,50 @@ private Session loadSession(String id)
+    public void put(String id, Session session) throws Exception
+    {
+        release(id,session);
+    }
```
This can be moved to the interface as the `default` implementation.

sbordet commented on jetty-server/src/main/java/org/eclipse/jetty/server/session/SessionCache.java:
```diff
@@ -118,6 +118,16 @@ default Session renewSessionId(String oldId, String newId, String oldExtendedId,
+     * @param session
+     * @throws Exception
+     */
+    void add(String id, Session session) throws Exception;
```
Isn't this a breaking change? `SessionCache` seems to be an interface that people would implement, no?

sbordet commented on jetty-server/src/main/java/org/eclipse/jetty/server/session/SessionCache.java:
```diff
@@ -138,9 +148,24 @@ default Session renewSessionId(String oldId, String newId, String oldExtendedId,
+     */
+    void release(String id, Session session) throws Exception;
+    
+    
```
Ditto with `add()` and too many blank lines.

sbordet commented on jetty-server/src/main/java/org/eclipse/jetty/server/session/SessionHandler.java:
```diff
@@ -1175,6 +1114,20 @@ public void renewSessionId(String oldId, String oldExtendedId, String newId, Str
+                    LOG.warn(e);
+                }
+            }
+        }
```
This is weird to me. Javadocs for `release()` say that the method should be called once a request is finished with a session.
It does not strike me that the request is finished with this session before existing this method. What am I missing?

janbartel replied:
The javadoc on line 1103 says that the request reference count will be incremented by the call to sessionCache.renewSessionId;  the finally clause makes sure that this increment is decremented again. 

You have to consider the case where the session id exists on multiple different contexts, and thus must all be renewed when a request invokes changeId method.  There may be no active requests in any of the other contexts, thus sessioncache for each context needs to load in the session if its not already present - this has to increment a reference count. If we did not then decrement the count when we've finished changing the session id, the session would sit forever in the sessioncaches of these other contexts with ref count =1.  By always incrementing and decrementing for change of session id we maintain the correct behaviour of the caches.

sbordet commented on tests/test-sessions/test-sessions-common/src/test/java/org/eclipse/jetty/server/session/AsyncTest.java:
```diff
@@ -216,6 +210,62 @@ public void testSessionWithCrossContextAsync() throws Exception
+    @Test
+    public void testSessionCreatedBeforeDispatch() throws Exception
+    {
+
```
Unnecessary empty line.

janbartel replied:
But I like it.

sbordet commented on tests/test-sessions/test-sessions-common/src/test/java/org/eclipse/jetty/server/session/AsyncTest.java:
```diff
@@ -216,6 +210,62 @@ public void testSessionWithCrossContextAsync() throws Exception
+        ServletHolder holder = new ServletHolder(servlet);
+        contextHandler.addServlet(holder, mapping);
+
+
```
Unnecessary empty lines.

janbartel replied:
But I like them.

sbordet commented on tests/test-sessions/test-sessions-common/src/test/java/org/eclipse/jetty/server/session/AsyncTest.java:
```diff
@@ -216,6 +210,62 @@ public void testSessionWithCrossContextAsync() throws Exception
+        }
+        finally
+        {
+            server.stop();
```
Also the client must be stopped.

janbartel replied:
OK.

sbordet commented on tests/test-sessions/test-sessions-common/src/test/java/org/eclipse/jetty/server/session/AsyncTest.java:
```diff
@@ -216,6 +210,62 @@ public void testSessionWithCrossContextAsync() throws Exception
+    
+    
+    
+    
```
_Many_ unnecessary empty lines!

janbartel replied:
Yeah, but I like them.

sbordet commented:
I'm also slightly worried about locking. 
The code references many components, each of which has its own lock.
Are we always sure that the lock ordering is the same (i.e. A grabs a_lock then calls B which grabs b_lock) and we never end up in a situation where the order is inverted (B grabs b_lock and then calls A which grabs a_lock)? Because otherwise it's a possible deadlock.

janbartel commented on jetty-server/src/main/java/org/eclipse/jetty/server/Request.java:
```diff
@@ -367,6 +368,39 @@ public void addEventListener(final EventListener listener)
+        if (s == null)
+            return;
+        
+        Session session = Session.class.cast(s);
```
OK.

janbartel commented on jetty-server/src/main/java/org/eclipse/jetty/server/Request.java:
```diff
@@ -367,6 +368,39 @@ public void addEventListener(final EventListener listener)
+     *
+     * @param s the session
+     */
+    private void leaveSession(final HttpSession s)
```
OK

janbartel commented on jetty-server/src/main/java/org/eclipse/jetty/server/HttpChannel.java:
```diff
@@ -1046,6 +1046,7 @@ private void notifyResponseFailure(Request request, Throwable failure)
 
     private void notifyComplete(Request request)
     {
+        request.onCompleted();
```
Fine, I'll move it to HttpChannel.onCompleted.

janbartel commented:
@sbordet re deadlocking:  I believe that the locking is sound but we can discuss on hangout.

janbartel commented:
Added comment on issue #3913 describing race condition that was fixed.

During fixing of this issue I also discovered and fixed these 2 problems:

1. If a request creates a session in eg contextA, forwards to a chain of other contexts, then forwards back to contextA, we will not be able to find the newly created session because we only retain the previous session & SessionHandler. Note that forwarding to only 1 other context before forwarding back to contextA the problem does not occur, because the previous context is contextA.

2. Async completion was not being handled correctly: we were registering an AsyncListener if asyncStarted == true when the SessionHandler exited. However the AsyncListener will only clean up a session if it was created in the same context for which async was started: if the request visits other contexts and creates sessions in them before async completion, none of those sessions will be cleaned up correctly.

For a full list of session use cases considered and verified for the fix, see the attached file.

[httpsession-fix.txt](https://github.com/eclipse/jetty.project/files/3491086/httpsession-fix.txt)
